### PR TITLE
initial handling of DynamicDependency in IDE0051

### PR DESCRIPTION
**Pull request by Youssef1313**
Contributes to #48206
We might want to keep the issue open if we want to track more advanced scenarios.

### CODE REVIEW

**Comment by Youssef1313** [comment]
This is definitely not ready to merge, but needs to get some feedback on whether it's on the right way or not.
@mavasani

File: src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -539,6 +549,20 @@ private ArrayBuilder<string> GetDebuggerDisplayAttributeArguments(INamedTypeSymb
                 return builder;
             }
 
+            private ArrayBuilder<string> GetDynamicDependencyAttributeArguments(INamedTypeSymbol namedTypeSymbol)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs:552]: how do you know this type isn't null?

File: src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -432,6 +437,11 @@ private void OnSymbolEnd(SymbolAnalysisContext symbolEndContext, bool hasUnsuppo
                                 continue;
                             }
 
+                            if (dynamicDependencyAttributeArguments.Any(signature => signature.Contains(member.Name)))
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs:440]: yeah, that seems like a non-good check.  i think we'd have to generate the actual doc string for something and compare it to the attribute arg.

File: src/Analyzers/CSharp/Tests/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 3]
```diff
@@ -2641,5 +2710,27 @@ public async Task UsedPrivateExtensionMethod()
...
+            var code = @"
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+
+class Program
+{
+    static void Main() => new Program().Bar();
+
+    [DynamicDependency(""Foo"")]
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/RemoveUnusedMembers/RemoveUnusedMembersTests.cs:2727]: is it legal to just name 'Foo' without things like `Foo()`, that seems problematic as we'll have to have some logic to figure out how to map these sorts of things to overload groups.

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/RemoveUnusedMembers/RemoveUnusedMembersTests.cs:2727]: is there a runtime spec for this anywhere that we can look at?

Comment (Youssef1313) [reply src/Analyzers/CSharp/Tests/RemoveUnusedMembers/RemoveUnusedMembersTests.cs:2727]: @CyrusNajmabadi I'm not sure if there is a spec for this somewhere. But basically writing just the name means **all overloads**.

There are examples for pointing to a **specific** attribute in @stephentoub's comment in the linked issue.

File: src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs:440]: @CyrusNajmabadi Yeah the current check is absolutely a bug. I was doing this quickly as an initial implementation that's not ready yet.

File: src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs (hunk 1 above)
Comment (Youssef1313) [reply src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs:552]: @CyrusNajmabadi The value passed comes from symbolEndContext.Symbol, which should be non-null.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi One of the approaches that can come to my mind involves the following steps:

1. Before doing anything, Get **all** members in the compilation using `CompilationStartContext.Compilation.GetSymbolsWithName((string _) => true, SymbolFilter.Member, cancellationToken)`.

    I'm expecting this call to be costly. Not sure if this is something you'd want to do.

2. For each of the returned symbols, find the DynamicDependencyAttribute and get the string passed to it.

3. Call `DocumentationCommentId.GetSymbolsForDeclarationId(dynamicDependencyArgument, compilation)`

4. Branch based on the result:

    1. If the result have 1 item only, add the symbol to a collection that holds the dynamic dependencies symbols
    2. If the result have more than an item, find the one contained in the same type as the symbol that declared the attribute.

Problems with this approach:

1. Cost.
2. Will still need more tweaks to handle things correctly.

Any ideas? @CyrusNajmabadi @mavasani

**Comment by mavasani** [comment]
Agree with @CyrusNajmabadi. Let us first start by getting the complete spec for this attribute before attempting to design how to decode it.

**Comment by Youssef1313** [comment]
> 
> 
> Agree with @CyrusNajmabadi. Let us first start by getting the complete spec for this attribute before attempting to design how to decode it.

Should I close until this happens?

**Comment by CyrusNajmabadi** [comment]
That, or move to draft. But closing send reasonable to help with clutter.

**Pull request closed without merging**